src/connector/src/source/base.rs

```diff
@@ -180,6 +185,26 @@ pub trait SplitMetaData: Sized {
 /// split readers.
 pub type ConnectorState = Option<Vec<SplitImpl>>;
 
+/// Used for acquiring the generated data for [`spawn_data_generation`].
+pub type DataGenerationReceiver = mpsc::Receiver<Result<Vec<SourceMessage>>>;
+
+/// Spawn a new **thread** to run the data generator, returns a channel receiver for acquiring the
+/// generated data. This is used for the [`DatagenSplitReader`] and [`NexmarkSplitReader`] in case
+/// that they are CPU intensive and may block the streaming actors.
+pub fn spawn_data_generation(
+    mut generator_next: impl FnMut() -> Result<Vec<SourceMessage>> + Send + 'static,
+) -> DataGenerationReceiver {
+    let (generation_tx, generation_rx) = mpsc::channel(4);
+    std::thread::spawn(move || loop {
```

Comment:
why need a os thread rather than a tokio thread

Reply:
If the data generator task is CPU intensive, it'll rarely yield to the runtime, which causes the streaming actors not to be scheduled at reasonable times. For example, when `split number = tokio worker thread number = parallel unit number`, we confirmed that the nexmark with interval `0ns` would immediately be stuck on compute nodes.

We may use multiple runtimes for different tasks on compute nodes in the future. For now, considering that the data generation is a kind of external thing, I think spawning a system thread for it makes some sense.

Reply:
However in deterministic test, external system thread is not allowed, otherwise determinism can not be achieved.
So you may have to maintain the original version as a special treatment for deterministic test. 🙃

Reply:
Does `spawn_blocking` work with Masson? 🤔

Reply:
No. since everything is scheduled on a single thread, it won't work if any function blocks the thread.🥵

Reply:
I've changed to this. It should work now? 🥺

https://github.com/singularity-data/risingwave/blob/d20848eab8c331122ba6392ecbda0c9c0afe8dd7/src/connector/src/source/base.rs#L211-L226